examples/quarternions.dx

```diff
@@ -0,0 +1,386 @@
+
+:html show3D $ one * rot * rot * rot
+
+:html show3D $ one * rot * rot * rot * rot
+
+' All the way around!
+
+
+rot2 = euler_to_quat (E [pi / 16.0, pi/8.0, 0.0]) 
+
+:html show3D $ one * rot2
+
+' ## Quaternion Conjugate
+The *conjugate* of a quaternion \(q\) is:
+
+' $$q^{*} = a-b\textrm{i}-c\textrm{j}-d\textrm{k}$$
+
+
+def conj ((Q {x, y, z, w}):Quaternion) : Quaternion =
+    Q {x=-x, y=-y, z=-z, w=w}
+
+
+' ## Quaternion Inverse
+
+' The *inverse* of a quaternion \(q\), is the conjugate divided by the magnitude squared:
+$$q^{-1} = \frac{q^{*}}{||q||^{2}}$$
+
+
+def inv (q:Quaternion) : Quaternion =
+    m = magnitude $ as_vec q
+    1.0 / (m * m) .* conj q
+
+' For unit quaternions, the conjugate is equal to the inverse.
+Multiplying a quaternion by its inverse results in the identity quaternion:
+
+' $$q.q^{-1} = (0, 0, 0, 1)$$
+
+
+' Let's try it out
+
+q' = from_vec $ normalize $ as_vec $ Q { x=1.0, y=2.0, z=0.5, w=10.0}
+
+
+:html show3D $ one 
+
+:html show3D $ one * q'
+
+:html show3D $ one * (inv q')
+
+:html show3D $ one * q' * (inv q')
+
+
+
+
+
+' ## Quaternion difference
+
+' The difference of two quaternions \(q_{1}\) and \(q_{2}\) is another quaternion \(q_{3}\) that rotates from \(q_{1}\) to \(q_{2}\):
+
+' $$q_{3} = q_{1}^{-1}.q_{2}$$
+
+difference = \ a b . (inv a) * b
+
+rot_right = euler_to_quat (E [0.0, pi/8.0, 0.0])
+:html show3D $ rot_right
+
+rot_right_down = euler_to_quat (E [pi/16.0, pi/8.0, 0.0]) 
+:html show3D $ rot_right_down
+
+
+rot_down = difference rot_right_down rot_right
+:html show3D $ rot_down
+
+
+' ## Quaternion Exp and Log
+
+' The exponential and the logarithm of a quaternion won't be very useful by themselves, but we will use them to compute other functions later.
+
+' Given a quaternion \(q = (x,y,z,w)\) and its vector part \(v = (x,y,z)\), the exponential of that quaternion is also a quaternion, and it's given by this formula:
+
+' $$\exp(q) = \exp(w)\begin{pmatrix}
+\frac{v_{x}}{||v||}\sin(||v||)\\\\
+\frac{v_{y}}{||v||}\sin(||v||)\\\\
+\frac{v_{z}}{||v||}\sin(||v||)\\\\
+\cos(||v||)
+\end{pmatrix}$$
+
+
+def q_exp (q:Quaternion) : Quaternion =
+    (Q {x,y,z,w}) = q
+    v = [x, y, z]
+    m = magnitude v
+    [x', y', z'] = normalize v
+    sin_v = sin m
+    exp_w = exp w
+    Q {x = x' * sin_v * exp_w,
+      y = y' * sin_v * exp_w,
+      z = z' * sin_v * exp_w,
+      w = cos(m) * exp_w,
+    }
+
+' Tangeant: to implement quaternion logs we need arccos. We don't have this yet in Dex
+  so this is an approximation (https://developer.download.nvidia.com/cg/acos.html)
+
+def acos (x:Float) : Float = 
+  negate = select (x < 0.0) 1.0 0.0
+  x = abs x
+  withState (-0.0187293) \ret .
+    ret := (get ret) * x
+    ret := (get ret) + 0.0742610
+    ret := (get ret) * x
+    ret := (get ret) - 0.2121144
+    ret := (get ret) * x
+    ret := (get ret) + 1.5707288
+    ret := (get ret) * (sqrt (1.0-x))
+    ret := (get ret) - 2.0 * negate * (get ret)
+    negate * 3.14159265358979 + (get ret)
```

Comment:
There's no looping so could you please write this out as immutable code? We haven't implemented simplification for simple state blocks like this one